### PR TITLE
[5.2] Make "required" validation rule work even if array key is missing

### PR DESCRIPTION
**Pull request by themsaid**
Currently this validation will pass although there's a person with no email.

``` php
$data = [
    'persons' => [
        [
            'name' => 'Jon Snow',
            'email' => 'jon@thewall.com',
        ],
        [
            'name' => 'Arya Stark',
        ]        
    ]
];

Validator::make($data, [
    'persons.*.email' => 'required'
]);
```

This is because `Arr::dot($this->data)` inside `Illuminate\Validation\Validator::each` flattens _only_ the existing keys and we pass them _only_ to the checks.

The proposed fix will create a null-value key if the key is missing so that it enters the validation checks and fail indicating that it's required but missing.

This fix is only taking care of the `required` rule, however the other `required_with`, `required_if`, `required_unless` and what's alike are addressed in the pull request https://github.com/laravel/framework/pull/11927

### CODE REVIEW

File: src/Illuminate/Validation/Validator.php [hunk 1]
```diff
@@ -274,6 +274,16 @@ public function each($attribute, $rules)
...
         $data = Arr::dot($this->data);
 
+        // We will check if the attribute is a nested key, if yes we bring the dot
+        // address of that key as well as the key name itself, the matches will
+        // be use to fill the the key with null if it's not provided for the
+        // "required" rule to work effectively.
+        preg_match("/(.*)\.\*\.([^\.\*]*)$/", $attribute, $matches);
+        if ($matches) {
+            $rootArraySegments = explode('.*.', $matches[1]);
+            $this->fillMissingArrayKeys($data, $matches[2], $this->data, $rootArraySegments, $rootArraySegments[count($rootArraySegments) - 1]);
```
Comment (sebdesign) [inline src/Illuminate/Validation/Validator.php]:
```php
$lastSegment = last($rootArraySegments);
$this->fillMissingArrayKeys($data, $matches[2], $this->data, $rootArraySegments, $lastSegment);
```

File: src/Illuminate/Validation/Validator.php [hunk 2]
```diff
@@ -286,6 +296,40 @@ public function each($attribute, $rules)
...
+                    continue;
+                }
+
+                $keysArray[] = $key;
+                $this->fillMissingArrayKeys($originalData, $field, $levelValues, $segments, $lastSegment, $keysArray);
+            } else {
+                foreach ($levelValues as $i => $rootValue) {
+                    if (! isset($rootValue[$field])) {
+                        $initialAddress = implode('.', $keysArray);
+                        $originalData[($initialAddress ? $initialAddress.'.' : '').($lastSegment.'.'.$i.'.'.$field)] = null;
```
Comment (sebdesign) [inline src/Illuminate/Validation/Validator.php]:
```php
$keys = array_merge($keysArray, [$lastSegment, $i, $field]);
$originalData[implode('.', $keys)] = null;
```

File: src/Illuminate/Validation/Validator.php [hunk 3]
```diff
@@ -274,8 +274,22 @@ public function each($attribute, $rules)
     {
         $data = Arr::dot($this->data);
 
+        // We will check if the attribute is a nested key, if yes we bring the dot
+        // address of that key as well as the key name itself, the matches will
+        // be use to fill the the key with null if it's not provided for the
```
Comment (phroggyy) [inline src/Illuminate/Validation/Validator.php]: the the

Comment (themsaid) [inline src/Illuminate/Validation/Validator.php]: @phroggyy fixed, thank you :)

**Comment by KoenRijpstra** [comment]
Validating an empty array returns the following error:
```
[2016-02-10 21:08:17] production.ERROR: exception 'ErrorException' with message 'Invalid argument supplied for foreach()' in /home/vagrant/Code/test/vendor/laravel/framework/src/Illuminate/Validation/Validator.php:304
```

Code:
```
$data = [];

Validator::make($data, [
    'persons.*.email' => 'required'
]);
```

**Comment by themsaid** [comment]
@KoenRijpstra That was fixed in: https://github.com/laravel/framework/pull/12243

**Comment by KoenRijpstra** [comment]
@themsaid Thanks! :+1:

**Comment by lino10** [comment]
@KoenRijpstra Hi, I'm just update to 5.2.20 but the bug is still there. So I replaced the Validator.php that you changed, but the bug is till there too. The 'required' is not kicking if I don't input any item_id.

My rule is like this
```
'item_id.*' => 'required|integer',
```

Html
```
<input name="item_id[]" type="hidden">
```

**Comment by themsaid** [comment]
@lino10 can you please write a failing test?

**Comment by lino10** [comment]
@themsaid I have a workaround, I don't know if it's the right way or not..
I added additional rule without the asterisk
```
'item_id' => 'required',
```

Really not elegant :cry:

**Comment by themsaid** [comment]
Can you please `dd($request->all());` after submitting the form with the failing validation to better investigate the case?

**Comment by lino10** [comment]
@themsaid The problem is the validation is not failing when it's supposed to fail.
I use the new FormRequest validation.

I have a form which you can input item dynamically, all the item ids are stored in a field(in my case hidden field). If user input new item, new field will be generated with the item id. 

```
<form>
<input type="text" name="room">
<input type="text" name="facility">
<!-- the dynamic field will be generated here -->
<input name="item_id[]" type="hidden" value="new item id"> <!-- example generated field -->
</form>
```

So if the user did not input any item and just submit the form, the validation won't trigger. I think it's because Laravel does not detect any array field with that name since it's not generated?

I hope you can understand my explanation. If this is the way Laravel works, then just ignore my issue :)

**Pull request merged**